File: benchmarks/array_processing_benchmark.ts
// TypeScript Array Processing Benchmark
// Equivalent to Cypescript benchmark for performance comparison

console.log("=== TypeScript Array Processing Benchmark ===");

// Configuration
const arraySize: number = 1000000; // 1 million elements
console.log("Array size: 1,000,000 elements");

// Create large array
console.log("Creating large array...");
const benchmarkArray: number[] = [];
for (let i: number = 1; i <= 100; i++) {
    benchmarkArray.push(i);
}

const actualSize: number = benchmarkArray.length;
console.log(`Actual benchmark array size: ${actualSize}`);

// Benchmark 1: Array Sum
console.log("Benchmark 1: Array Sum");
const iterations: number = 100000; // Repeat operation many times
console.log("Iterations: 100,000");

let sumTotal: number = 0;
console.log("Starting sum benchmark...");
const sumStart: number = performance.now();

for (let iter: number = 0; iter < iterations; iter++) {
    let currentSum: number = 0;
    
    // Sum all elements in array
    for (let j: number = 0; j < actualSize; j++) {
        currentSum += benchmarkArray[j];
    }
    
    sumTotal += currentSum;
}

const sumEnd: number = performance.now();
const sumTime: number = sumEnd - sumStart;
console.log("Sum benchmark complete");
console.log(`Total sum across all iterations: ${sumTotal}`);
console.log(`Sum benchmark time: ${sumTime.toFixed(2)}ms`);

// Benchmark 2: Array Maximum
console.log("Benchmark 2: Array Maximum");
const maxIterations: number = 100000;
console.log("Iterations: 100,000");

let maxTotal: number = 0;
console.log("Starting max benchmark...");
const maxStart: number = performance.now();

for (let maxIter: number = 0; maxIter < maxIterations; maxIter++) {
    let currentMax: number = benchmarkArray[0];
    
    // Find maximum element
    for (let k: number = 1; k < actualSize; k++) {
        if (benchmarkArray[k] > currentMax) {
            currentMax = benchmarkArray[k];
        }
    }
    
    maxTotal += currentMax;
}

const maxEnd: number = performance.now();
const maxTime: number = maxEnd - maxStart;
console.log("Max benchmark complete");
console.log(`Total max across all iterations: ${maxTotal}`);
console.log(`Max benchmark time: ${maxTime.toFixed(2)}ms`);

// Benchmark 3: Array Search
console.log("Benchmark 3: Array Search");
const searchIterations: number = 100000;
const searchTarget: number = 75;
console.log("Iterations: 100,000");
console.log(`Search target: ${searchTarget}`);

let foundCount: number = 0;
console.log("Starting search benchmark...");
const searchStart: number = performance.now();

for (let searchIter: number = 0; searchIter < searchIterations; searchIter++) {
    let found: boolean = false;
    
    // Linear search for target
    for (let l: number = 0; l < actualSize; l++) {
        if (benchmarkArray[l] === searchTarget) {
            found = true;
            break; // TypeScript can break early
        }
    }
    
    if (found) {
        foundCount++;
    }
}

const searchEnd: number = performance.now();
const searchTime: number = searchEnd - searchStart;
console.log("Search benchmark complete");
console.log(`Found target in iterations: ${foundCount}`);
console.log(`Search benchmark time: ${searchTime.toFixed(2)}ms`);

// Benchmark 4: Array Modification
console.log("Benchmark 4: Array Modification");
const modifyIterations: number = 50000; // Fewer iterations for modification
console.log("Iterations: 50,000");

// Create a copy for modification
const modifyArray: number[] = [...benchmarkArray];

console.log("Starting modification benchmark...");
const modifyStart: number = performance.now();

for (let modifyIter: number = 0; modifyIter < modifyIterations; modifyIter++) {
    // Double all elements
    for (let m: number = 0; m < actualSize; m++) {
        modifyArray[m] *= 2;
    }
    
    // Reset array for next iteration
    for (let n: number = 0; n < actualSize; n++) {
        modifyArray[n] /= 2;
    }
}

const modifyEnd: number = performance.now();
const modifyTime: number = modifyEnd - modifyStart;
console.log("Modification benchmark complete");
console.log(`Modification benchmark time: ${modifyTime.toFixed(2)}ms`);

// Summary
console.log("=== TypeScript Benchmark Summary ===");
console.log(`Sum benchmark: ${sumTime.toFixed(2)}ms`);
console.log(`Max benchmark: ${maxTime.toFixed(2)}ms`);
console.log(`Search benchmark: ${searchTime.toFixed(2)}ms`);
console.log(`Modification benchmark: ${modifyTime.toFixed(2)}ms`);
console.log(`Total time: ${(sumTime + maxTime + searchTime + modifyTime).toFixed(2)}ms`);
console.log("=== All Benchmarks Complete ===");

File: benchmarks/array_processing_benchmark.py
#!/usr/bin/env python3
# Python Array Processing Benchmark
# Equivalent to Cypescript array_processing_benchmark.csc

import time

print("=== Python Array Processing Benchmark ===")

# Configuration
array_size = 1000000  # 1 million elements
print(f"Array size: {array_size:,} elements")

# Create large array
print("Creating large array...")
test_array = list(range(1, 101))  # 1-100 for comparison with Cypescript
actual_size = len(test_array)
print(f"Actual benchmark array size: {actual_size}")

# Benchmark 1: Array Sum
print("Benchmark 1: Array Sum")
iterations = 100000  # 100,000 iterations
print(f"Iterations: {iterations:,}")

sum_start = time.perf_counter()
sum_total = 0
for iter_count in range(iterations):
    current_sum = sum(test_array)
    sum_total += current_sum
sum_end = time.perf_counter()
sum_time = (sum_end - sum_start) * 1000  # Convert to milliseconds

print("Sum benchmark complete")
print(f"Total sum across all iterations: {sum_total}")
print(f"Sum benchmark time: {sum_time:.2f}ms")

# Benchmark 2: Array Maximum
print("Benchmark 2: Array Maximum")
max_iterations = 100000
print(f"Iterations: {max_iterations:,}")

max_start = time.perf_counter()
max_total = 0
for iter_count in range(max_iterations):
    current_max = max(test_array)
    max_total += current_max
max_end = time.perf_counter()
max_time = (max_end - max_start) * 1000

print("Max benchmark complete")
print(f"Total max across all iterations: {max_total}")
print(f"Max benchmark time: {max_time:.2f}ms")

# Benchmark 3: Array Search
print("Benchmark 3: Array Search")
search_iterations = 100000
search_target = 75
print(f"Iterations: {search_iterations:,}")
print(f"Search target: {search_target}")

search_start = time.perf_counter()
found_count = 0
for iter_count in range(search_iterations):
    found = search_target in test_array
    if found:
        found_count += 1
search_end = time.perf_counter()
search_time = (search_end - search_start) * 1000

print("Search benchmark complete")
print(f"Found target in iterations: {found_count}")
print(f"Search benchmark time: {search_time:.2f}ms")

# Benchmark 4: Array Modification
print("Benchmark 4: Array Modification")
modify_iterations = 50000  # Fewer iterations for modification
print(f"Iterations: {modify_iterations:,}")

# Create a copy for modification
modify_array = test_array.copy()

modify_start = time.perf_counter()
for iter_count in range(modify_iterations):
    # Double all elements
    for i in range(len(modify_array)):
        modify_array[i] *= 2
    
    # Reset array for next iteration
    for i in range(len(modify_array)):
        modify_array[i] //= 2
modify_end = time.perf_counter()
modify_time = (modify_end - modify_start) * 1000

print("Modification benchmark complete")
print(f"Modification benchmark time: {modify_time:.2f}ms")

# Summary
total_time = sum_time + max_time + search_time + modify_time
print("=== Python Benchmark Summary ===")
print(f"Sum benchmark: {sum_time:.2f}ms")
print(f"Max benchmark: {max_time:.2f}ms")
print(f"Search benchmark: {search_time:.2f}ms")
print(f"Modification benchmark: {modify_time:.2f}ms")
print(f"Total time: {total_time:.2f}ms")
print("=== All Benchmarks Complete ===")
